src/org/jetbrains/java/decompiler/struct/StructContext.java

```diff
@@ -140,29 +140,20 @@ else if (filename.endsWith(".zip")) {
     }
   }
 
-  private FileSystem getZipFileSystem(File file) throws IOException {
-    try {
-      return zipFiles.computeIfAbsent(file, f -> {
-        URI uri = null;
-        try {
-          uri = new URI("jar:file", null, f.toURI().getPath(), null);
-          return FileSystems.newFileSystem(uri, Collections.emptyMap());
-        } catch (FileSystemAlreadyExistsException e) {
-          return FileSystems.getFileSystem(uri);
-        } catch (URISyntaxException e) {
-          throw new RuntimeException(e);
-        } catch (IOException e) {
-          throw new UncheckedIOException(e);
-        }
-      });
-    } catch (UncheckedIOException e) {
-      throw e.getCause();
-    }
-  }
-
   private void addArchive(String externalPath, File file, int type, boolean isOwn) throws IOException {
     DecompilerContext.getLogger().writeMessage("Adding Archive: " + file.getAbsolutePath(), Severity.INFO);
-    FileSystem fs = getZipFileSystem(file);
+    FileSystem fs;
+    try {
+      URI uri = new URI("jar:file", null, file.toURI().getPath(), null);
+      try {
+        fs = FileSystems.newFileSystem(uri, Collections.emptyMap());
+        toClose.add(fs);
+      } catch (FileSystemAlreadyExistsException e) {
+        fs = FileSystems.getFileSystem(uri);
+      }
```

Comment:
Slight issue: this doesn't work with multiple threads. One could create the new file system, add it to `toClose`; another thread uses the already open file system; the first thread closes it; the second thread tries to use the closed file system.

Reply:
Such a project would require a complex synchronization system for FileSystem closing and should just ensure it is open for the duration of decompilation. Similar code is used in stitch and tiny remapper